### ISSUE TITLE
Uncontrolled resource consumption issue with fuzzed sample

### ISSUE THREAD

**Issue by fgeek** [htacg/tidy-html5#459]
I found uncontrolled resource consumption [CWE-400](https://cwe.mitre.org/data/definitions/400.html) (also called resource exhaustion) issue with fuzzing tidy using american fuzzy lop version 2.35b. Tested in d2fc25259859abd8c7d0e29dd23e790d48167a2c. You can reproduce this case with:
> wget http://bugs.fi/media/afl/tidy-html5/3c4243596f495b64345f39d807ea3263bd34d4a3.html
> tidy-html5 3c4243596f495b64345f39d807ea3263bd34d4a3.html

Is this something that you consider relevant to be fixed in tidy-html5?

**Comment by geoffmcl** [comment]
@fgeek I tested the bugs.html, from our direct exchange, and now the above file, and **yes**, tidy gets stuck in a parsing loop... still in process of tracking it down exactly... and thanks for bringing it here, and adding a test file... that makes testing easy...

Being only a single process, I did not see `uncontrolled resource consumption`, `resource exhaustion`... the cpu usage rose, but capped at about 20-50%, saw no significant memory increase, and still to check stack space... but agree these are **bad** signs... 

S/w should **never** get `stuck` in a **forever** loop ;=))

And must look up some recent similar problems when parsing sort of random binary data streams, and what was decided, done there... your files certainly have interesting sequences...

I remember one option was to simply abort, and return error, if say zero(s) found, or other bytes, not part of the input charset, but I think in that case, a solution was found... again need a little time to re-read all that... maybe some others can remind me...

But, yes. in general this is a **bug** that we would like to **fix**... tidy just should not have a problem with any random input byte stream... any code, patches, suggestions welcome... thanks...

**Comment by fgeek** [comment]
My pleasure to help. I do not have the skills to fix this with patch, sorry. For the record we did not find any security issues while fuzzing the binary. Have a nice weekend.

**Comment by geoffmcl** [comment]
@fgeek some back reading, and found -

Issue #341 was about some allocated memory, but it was there that there was some discussion on aborting the parsing in certain binary file cases... related, but not the same...

There is another binary file mentioned in #432... but again different case...

Will get back to trying the exactly understand the sequence that puts tidy into a **fatal** tail spin... thanks...

**Comment by geoffmcl** [comment]
@fgeek ok, found a problem in some code I added circa 20150201, Issue #166, v.4.9.6... It needs to check every node, but with the current code, that can be tens, hundreds, if not thousand, exponentially growing, checking each node multiple, multiple times... and potentially consuming heavy stack space...

Have **fixed** the code, and now tidy has no problem with either of your samples... the patch is -

````patch
diff --git a/src/parser.c b/src/parser.c
index 4fb7a8c..30a9911 100644
--- a/src/parser.c
+++ b/src/parser.c
@@ -3851,7 +3851,15 @@ Bool TY_(FindNodeWithId)( Node *node, TidyTagId tid )
     {
         if (TagIsId(node,tid))
             return yes;
-        for (content = node->content; content; content = content->content)
+        /*\ 
+         *   Issue #459 - Under certain circumstances, with many nodes, this use of
+         *   'for (content = node->content; content; content = content->content)'
+         *   would produce a **forever** circle, or at least a very extended loop...
+         *   It is sufficient to test the content, if it exists,
+         *   to quickly iterate all nodes. Now all nodes are tested only once.
+        \*/ 
+        content = node->content;
+        if (content)
         {
             if (TY_(FindNodeWithId)(content,tid))
                 return yes;
diff --git a/version.txt b/version.txt
index 9c3783d..f9c3f12 100644
--- a/version.txt
+++ b/version.txt
@@ -1,2 +1,2 @@
-5.3.11
-2016.09.11
+5.3.12
+2016.10.14
````

With it, each node in the tree is now only checked once... should be very quick, and not consume much stack space... bug squashed?

I will try to get around to pushing this, after some more testing, but meantime maybe you could apply the patch, and re-test tidy v.5.3.12... thanks...

**Comment by fgeek** [comment]
Patch seems to fix this issue. Thank you for fast response and work. If I find some additional bugs I will create new issue to GH.

**Comment by geoffmcl** [comment]
@fgeek thanks for the `response` comment, but took me `weeks` to get to the push ;=((

Hope version 5.3.12 meets those needs...

Thanks again for the report...
